### PR TITLE
WIP Quick doc

### PR DESCRIPTION
**Pull request by LeoDaCoda**
Hey, this PR is in reference to an [issue](https://github.com/gitpython-developers/GitPython/issues/1605) I raised. I would like to add a quick start guide to act as another resource for developers, I planned the guide to be mainly code snippets as opposed to a text based tutorial.

I finished all the code and section headers I planned to add to the quick doc. I still want to add a few short sentences to improve readability. The tutorials includes how to do basic repo operations like init, clone, add, commit and push. Also shows how to iterate through the trees and print a text file.

Where possible I tried to follow the conventions the regular tutorial followed (ie having the code snippets inside of the test file) but in some cases I had to add code directly to the RST (I did this to avoid having unnecessary print statements to the test suite).

This is my first PR so forgive me if I did not follow propper conventions. Look forward to your feedback, Thanks!

### CODE REVIEW

**Comment by LeoDaCoda** [comment]
Also i got this warning from 
```sphinx-build -b html . _build/html```

checking consistency... /home/jay/PycharmProjects/GitPython-QuickStart-Doc/doc/source/quickstart.rst: WARNING: document isn't included in any toctree

Unsure how to fix this

**Review by Byron — CHANGES_REQUESTED** [review]
Thanks a lot for taking the time to figure out how to contribute to the documentation setup used in `GitPython`, it's absolutely awesome to see it improve after all this time.

The goal of the quick-start section is to show code people can copy-paste and as such it should represent the best `GitPython`, and python, have to offer. Thus I may have a lot of opinions in the following comments, but those are merely to get closer to that (probably unattainable) goal.

Generally, please avoid assigning variables that are only used once unless they truly help to explain what the expression means, and names like `fileX` don't seem to do that.

Besides that, I think this is an excellent basis for getting this out to people fast. What would help me is if you could attach a PDF of the rendered document in the comments so I don't have to go through the trouble of building it locally (I have not been able to do anything with python for quite some time and it's great pain for me to deal with that again).

Looking forward to push this PR to an early completion, hoping that you can keep improving the document as you gain more insights into `GitPython` yourself.

File: test/test_quick_doc.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+    @with_rw_directory
+    def test_init_repo_object(self, rw_dir):
+        path_to_dir = rw_dir
+
+        # [1-test_init_repo_object]
+        from git import Repo
+
+        repo = Repo.init(path_to_dir)  # git init path/to/dir
+        assert repo.__class__ is Repo  # Test to confirm repo was initialized
```
[the remark this answers is not in the document]
Comment (LeoDaCoda) [reply test/test_quick_doc.py:22]: Do you mean the assertion statement? If so, I agree in removing them, when I first read the test_docs.py file I saw a lot of assert statements and assumed the tests methods should be testing something. 

Or do you think the having the Repo.init()  is unnecessary?

File: test/test_quick_doc.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        # Outputs: <generator object Commit._iter_from_process_or_stream at 0x7fb66c186cf0>
+
+        # ![5-test_cloned_repo_object]
+
+        # [6-test_cloned_repo_object]
+        commits_for_file_generator = repo.iter_commits('--all', max_count=100, paths=file)
+        commits_for_file = [c for c in commits_for_file_generator]
+        commits_for_file
+
+        # Outputs: [<git.Commit "5076b368c97b01d83406ca095a301303da7f6fd4">,
```
[the remark this answers is not in the document]
Comment (LeoDaCoda) [reply test/test_quick_doc.py:82]: ok

File: test/test_quick_doc.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        from git import Repo
+        import git
+        # code to clone from url
+        # [1-test_cloned_repo_object]
+        repo_url = "https://github.com/LeoDaCoda/GitPython-TestFileSys.git"
+
+        try:
+            repo = Repo.clone_from(repo_url, local_dir)
+        except git.CommandError:
+            assert False, f"Invalid address {repo_url}"
```
[the remark this answers is not in the document]
Comment (LeoDaCoda) [reply test/test_quick_doc.py]: removed try/except statements

File: test/test_quick_doc.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+        # ![11-test_cloned_repo_object]
+
+        '''Trees and Blobs'''
+
+        # Latest commit tree
+        # [12-test_cloned_repo_object]
+        tree = repo.tree()
+        # ![12-test_cloned_repo_object]
+
+        # Previous commit tree
```
[the remark this answers is not in the document]
Comment (LeoDaCoda) [reply test/test_quick_doc.py:125]: thanks! updated

**Comment by LeoDaCoda** [comment]
Sorry for the late response I was a little busy this week. I think I addressed all your comments and I attached the pdf output. I asked gpt to write a 1-2 sentence intro, but to me it still sounds a little wonky. 

I intended the quickstart`s audience to be developers that want to just code/experiment and do not want to comb through the entire tutorial and stackoverflow posts. Do you think the tutorial is missing anything
[Quickstar-print.pdf](https://github.com/gitpython-developers/GitPython/files/12035856/Quickstar-print.pdf)

File: test/test_quick_doc.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Byron) [reply test/test_quick_doc.py]: Removing the `assert` is the way to go here, if something would go wrong an exception would have been thrown.

**Comment by LeoDaCoda** [comment]
Hey Byron! I made some changes to make the document more readable like putting the equivalent bash in a comment instead as a sub-header. I also some made general changes to the structure and the order of the code-snippets. 

I attached the pdf print version.
[Quickstar-print.pdf](https://github.com/gitpython-developers/GitPython/files/12076042/Quickstar-print.pdf)

File: test/test_quick_doc.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+        # Outputs: <generator object Commit._iter_from_process_or_stream at 0x7fb66c186cf0>
+
+        # ![5-test_cloned_repo_object]
+
+        # [6-test_cloned_repo_object]
+        commits_for_file_generator = repo.iter_commits(all=True, max_count=10, paths=update_file)
+        commits_for_file = [c for c in commits_for_file_generator]
+        commits_for_file
+
+        # Outputs: [<git.Commit "SHA1-HEX_HASH-2">,
```
Comment (Byron) [inline test/test_quick_doc.py:79]:
```suggestion
        # Outputs: [<git.Commit "SHA1-HEX-HASH">,
```

File: test/test_quick_doc.py [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+        # <git.Commit "SHA1-HEX-HASH-2">]
```
Comment (Byron) [inline test/test_quick_doc.py:80]:
```suggestion
        # <git.Commit "SHA1-HEX-HASH">]
```

File: test/test_quick_doc.py [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+        # ----| dir1 / file2.txt, blob
+        # | file4.txt, blob
+        # # ![16-test_cloned_repo_object]
+
+        # Printing text files
+        # [17-test_cloned_repo_object]
+        print_file = 'dir1/file2.txt'
+        tree[print_file]  # the head commit tree
+
+        # Output <git.Blob "SHA1-HEX-HASH-1">
```
Comment (Byron) [inline test/test_quick_doc.py:212]:
```suggestion
        # Output <git.Blob "SHA1-HEX-HASH">
```

**Review by Byron — CHANGES_REQUESTED** [review]
Thanks again! I have a feeling that not all of my comments have been applied (I am looking at you, `.` ;)), but with the structural changes it doesn't strike me as problem anymore.

Once you have applied the last remaining changes to unify the depiction of hashes, this PR is good to go. Note that the idea there was to not let it appear they are the same by adding an incremental number at their tail, so now I go for the smallest common denominator and remove the numbering instead so they are at least consistent.

**Comment by LeoDaCoda** [comment]
At first I didn't understand what you meant by the `-1` and `-2` but after you explained it it made more sense. I tried to follow this convention: blob and tree would not have an appended number, but commits would be numbered in chronological order.

I also wanted to link some sections of the full tutorial in the relevant sections of the quick start but I ran into a problem. Sphinx requires references to follow this convention: `:ref:\`Remotes <handling-remotes>\``

**Comment by Byron** [comment]
Great, thanks for all the fixes!

Regarding the subsection labels, feel free to add them in the next PR :). Let's get this merged!

**Pull request merged**